src/typeconversions.jl

```diff
@@ -60,6 +60,16 @@ function datatype(str::VLen{C}) where {C<:CharType}
     Datatype(API.h5t_vlen_create(type_id))
 end
 
+# Compound types
+datatype(x::AbstractArray{T}) where {T} = HDF5.datatype(x, Val(isstructtype(T)))
+function datatype(x::AbstractArray{T}, isstruct::Val{true}) where {T}
+    dtype = API.h5t_create(API.H5T_COMPOUND, sizeof(T))
+    for (idx, fn) in enumerate(fieldnames(T))
+        API.h5t_insert(dtype, fn, fieldoffset(T, idx), datatype(fieldtype(T, idx)))
+    end
+    Datatype(dtype)
+end
```

Comment:
```suggestion

datatype(::T) where T = Datatype(hdf5_type_id(T), false)
datatype(::Type{T}) where T = Datatype(hdf5_type_id(T), false)
datatype(x::AbstractArray{T}) where T = Datatype(hdf5_type_id(T), false)


# Move closer to where hdf5_type_id is defined
hdf5_type_id(::Type{T}) where T = hdf5_type_id(T, Val(isstructtype(T)))
function hdf5_type_id(::Type{T}, isstruct::Val{true}) where T
    dtype = API.h5t_create(API.H5T_COMPOUND, sizeof(T))
    for (idx, fn) in enumerate(fieldnames(T))
        API.h5t_insert(dtype, fn, fieldoffset(T, idx), hdf5_type_id(fieldtype(T, idx)))
    end
    return dtype
end
# 
hdf5_type_id(::Type{T}, isstruct::Val{false}) where T = MethodError(hdf5_type_id, (T, isstruct))
```

How about this. Some of this logic belongs at the `hdf5_type_id` level whereas `datatype` is more of a frontend for wrapping the hdf5 type id as `Datatype`

Reply:
OK now I start to grasp. This `hdf5_type_id` was the key. I tried something similar but always ended up in infinite recursion...